### DISCUSSION TITLE
how to create wrapper for c/cpp structure member holding pointer to another structure

### DISCUSSION THREAD

**Discussion by soundarsurya** [pybind/pybind11#4036] [Q&A]
Hi,

I want to write python wrapper API, in my case i want to pass structure pointer as API argument. In real case it is more complex.
Attached my sample code implementation

[py_wrapper_impl.zip](https://github.com/pybind/pybind11/files/9007834/py_wrapper_impl.zip)

**Readme**
api.cpp -> user exposed api
core.cpp -> middle level api
implemenmtation.cpp -> actually implementation of function

**build**
rm -rf build
mkdir build && cd build
cmake ..
make

Kindly help to create wrapper for my case asap.

**Comment by jiwaszki** [comment]
@soundarsurya it would be better to create open Github repository. Passing around code (or anything in that matter) with zip files is not viewed as safe and secure.

**Comment by Nic30** [comment]
@soundarsurya 

1. There is an uncountable amount of project to work on. The lifespan is not sufficient to do all what is required. Considering that many people on github are working on project entirely for free, I would recommend to rephrase your command as a question? "I am new to a pybind11, I would like to bind some structs to a python, where do I start?"

2. There is "no difference" between the struct and class in pybind11 and py::class_ will work for both uncountable https://pybind11.readthedocs.io/en/stable/classes.html
    * The py::class_ can wrap value or a pointer, read more e.g. there https://pybind11.readthedocs.io/en/stable/advanced/smart_ptrs.html

**Reply by soundarsurya** [reply]
hi @Nic30 

Thanks for your comment. The below is what i'm trying. I have gone through the official document. But i'm not getting that how we can handle pointer in pybind11 wrapper. Below is the scenario in my code.

input_struct -> is a structure which contain float pointer and reference to another enum
i want to pass this 'input_struct' as pointer to api 'pyInit'

**/*header.hpp*/**
```
extern "C"{
typedef enum {
    E_CALCULATOR = 0,
    E_MEAUREMENTS
} option;

typedef struct {
    float *number1;
    float number2;
    option option_enum;
}input_struct;

int expose_init(const input_struct*);
int expose_operation(void);
}
```

**/*PyBind11_wrapper.cpp*/**
```
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>

namespace py = pybind11;

PYBIND11_MODULE(apiWrapper, module){
    module.doc() = "This is py wrapper for exposed APIs";
    py::class_<input_struct>(module, "input_struct")
    .def(py::init())
    .def("pyNum1", &input_struct::number1) # not sure about declaring pointer
    .def("pyNum2", &input_struct::number2)
    .def("pyOpti", &input_struct::option_enum); # not sure about declaring another struct/enum reference
    module.def("pyInit", &expose_init, "Init function"); # how we can pass the 'input_struct' pointer as argument
    module.def("PyOperation", &expose_operation, "Operation function");
}
```
**/*test_app.cpp*/**
Working Test application in cpp.
writing wrapper for the below two implemented APIs
expose_init -> PyInit
expose_operation -> PyOperation

```
#include <iostream>
#include "header.hpp"
using namespace std;
int main(){
    input_struct *inp_struct_ptr = new input_struct();
    float num = 20;
    inp_struct_ptr->number1 = &num;
    inp_struct_ptr->number2 = 100;

    cout <<"Calling expose_init" << endl;
    if(expose_init(inp_struct_ptr) < 0){
        cout << " Init calling failed!!! ";
        exit(-1);
    }
    cout <<"\nCalling expose_operation " << endl;
    if(expose_operation() < 0){
        cout << " Operation calling failed!!! ";
        exit(-1);
    }
    delete inp_struct_ptr;
    return 0;
}
```
